src/modules/sync/lib/ipc-handlers.ts

```diff
@@ -158,16 +166,94 @@ export async function pushArchive(
 		throw new Error( 'No token found' );
 	}
 
+	let isUploadingPaused = false;
+	const file = fs.createReadStream( archivePath );
+	const fileSize = fs.statSync( archivePath ).size;
+	const filename = path.basename( archivePath );
+
+	const cleanup = () => {
+		file.destroy();
+		file.close();
+	};
+
+	const attachmentId = await new Promise< string >( ( resolve, reject ) => {
+		const upload = new Upload( file, {
+			endpoint: `${ STUDIO_FILE_UPLOADS_ENDPOINT_BASE }/${ remoteSiteId }`,
+			chunkSize: STUDIO_FILE_UPLOADS_CHUNK_SIZE,
+			retryDelays: STUDIO_FILE_UPLOADS_RETRY_DELAYS,
+			overridePatchMethod: true,
+			removeFingerprintOnSuccess: true,
+			storeFingerprintForResuming: true,
+			headers: {
+				Authorization: `Bearer ${ token.accessToken }`,
+			},
+			metadata: {
+				filename,
+				filetype: 'application/gzip',
+			},
+			uploadSize: fileSize,
+			onBeforeRequest: ( req: HttpRequest ) => {
+				if ( req.getMethod() === 'HEAD' ) {
+					// @ts-expect-error We need to override the method to get the response headers.
+					req._method = 'GET';
+					req.setHeader( 'X-HTTP-Method-Override', 'HEAD' );
+				}
+			},
+			onError: ( error ) => {
+				cleanup();
+				console.error( '[TUS] Upload error', error );
+				reject( error );
+			},
+			onProgress: ( bytesSent, bytesTotal ) => {
+				if ( isUploadingPaused ) {
+					isUploadingPaused = false;
+					void sendIpcEventToRenderer( 'studio-file-upload-resumed', {
+						selectedSiteId: selectedSiteId,
+						remoteSiteId: remoteSiteId,
+					} );
+					console.log( '[TUS] Upload resumed' );
+				}
+
+				if ( bytesTotal > 0 ) {
+					const progress = ( bytesSent / bytesTotal ) * 100;
+					console.log( '[TUS] Upload progress: %s%%', progress.toFixed( 2 ) );
+				}
+			},
+			onSuccess: ( payload: { lastResponse: HttpResponse } ) => {
+				cleanup();
+
+				if ( ! payload.lastResponse ) {
+					console.error( 'Upload completed but no response received' );
+					reject( new Error( 'Upload completed but no response received' ) );
+					return;
+				}
+
+				const attachmentId = payload.lastResponse.getHeader( 'x-videopress-upload-media-id' );
+				if ( attachmentId ) {
+					resolve( attachmentId );
+				} else {
+					console.error( 'Upload completed but required header not found' );
+					reject( new Error( 'Upload completed but required header not found' ) );
+				}
+			},
+			onShouldRetry: ( error: Error ) => {
+				isUploadingPaused = true;
+				void sendIpcEventToRenderer( 'studio-file-upload-paused', {
+					selectedSiteId: selectedSiteId,
+					remoteSiteId: remoteSiteId,
+					error: error.message,
+				} );
+				console.error( 'Upload request error', error.message );
+				return true;
+			},
+		} );
+
+		upload.start();
+	} );
```

Comment:
```suggestion
	} ).finally( () => {
		file.destroy();
		file.close();
	} );

	const attachmentId = await attachmentPromise;
```
We recently encountered an issue with very similar logic in the CLI: the cleanup logic wouldn't run under specific circumstances. Your logic looks solid, but I still recommend this approach, as it's foolproof. Combine it with the following change on L174:

```ts
	const attachmentPromise = new Promise< string >( ( resolve, reject ) => {
```

Reply:
Also, combine it with removing the `cleanup` function

Reply:
I have applied your suggestion in [7aaf10d](https://github.com/Automattic/studio/pull/2212/commits/7aaf10d0180f36708d273c4690597b925e1a75da). I am curious, under what condition it doesn't run? Did notice anything specific?

Reply:
Your existing logic looked solid. I pointed it out because it was remarkably similar to a case that recently gave us a headache in the CLI project. Using a `finally` clause guarantees that cleanup logic runs at the right time, regardless of how the promise settles. Consider it future proofing 🙂

Reply:
Thanks for explaining it. It makes sense to use `finally` becuase no matter how promise settles, it will be always called.